File: core/weapons/manager.js
import { clamp } from "../utils/math.js";
import { isPromise } from "../utils/misc.js";
import PulseCanon from "./pulse-canon.js";
import GatlingGun from "./gatlingGun.js";
import HeavyRailGun from "./heavyRailGun.js";
import PlasmaCanon from "./plasmaCanon.js";
import Mine from "./mine.js";
import HomingMissile from "./HomingMissile.js";
import { keybinds } from "../events/keybind.js";
import { world } from "../world/world.js";
import { worldManager } from "../world/manager.js";

export default class WeaponManager {
    constructor() {
        this.#addKeybinds();
    }

    static weaponTypes = [PulseCanon, GatlingGun, HeavyRailGun, PlasmaCanon, HomingMissile, Mine];

    static weapons = [];

    static fire(Weapon, options, force = false) {
        WeaponManager.weapons.push(new Weapon(options, force));
    }

    static render() {
        for (const weapon of WeaponManager.weapons) {
            weapon.render();
        }
    }

    static update(t, dt) {
        for (const weapon of WeaponManager.weapons) {
            weapon.update(t, dt);
        }
    }

    static destroy(weapon) {
        const index = WeaponManager.weapons.indexOf(weapon);
        if (index > -1) {
            WeaponManager.weapons.splice(index, 1);
        }
    }


    #addKeybinds() {
        keybinds["q"] = () => {
            this.previousWeapon();
        }
        keybinds["e"] = () => {
            this.nextWeapon();
        }
    }

    nextWeapon() {
        const ship = worldManager.findAttachedShip();
        const currentWeaponId = WeaponManager.weaponTypes.findIndex(w => w === ship.weapon) + 1;

        const nextWeapon = WeaponManager.weaponTypes[currentWeaponId] || WeaponManager.weaponTypes[0];

        this.changeWeapon(nextWeapon);
    }

    previousWeapon() {
        const ship = worldManager.findAttachedShip();

        const currentWeaponId = WeaponManager.weaponTypes.findIndex(w => w === ship.weapon) - 1;
        const previousWeapon = WeaponManager.weaponTypes[currentWeaponId] || WeaponManager.weaponTypes[WeaponManager.weaponTypes.length - 1];

        this.changeWeapon(previousWeapon);
    }

    changeWeapon(weapon) {
        worldManager.findAttachedShip().weapon = weapon;
    }
}

// Todo: createWeapon pool
export const weaponManager = new WeaponManager();

File: void_front/game/core/weapons/manager.py
from __future__ import annotations

from game.core.events.keybind import keybinds
from game.core.world.manager import worldManager
import numpy as np


class WeaponManager:
    weapons: list = []
    weaponTypes: list = []

    def __init__(self):
        self._addKeybinds()

    @staticmethod
    def fire(Weapon, options, force=False):
        WeaponManager.weapons.append(Weapon(options, force))

    @staticmethod
    def render():
        for weapon in WeaponManager.weapons:
            weapon.render()

    @staticmethod
    def update(t, dt):
        for weapon in list(WeaponManager.weapons):
            weapon.update(t, dt)

    @staticmethod
    def destroy(weapon):
        if weapon in WeaponManager.weapons:
            WeaponManager.weapons.remove(weapon)

    def _addKeybinds(self):
        keybinds["q"] = self.previousWeapon
        keybinds["e"] = self.nextWeapon

    @staticmethod
    def _getWeaponTypes():
        from game.core.weapons.pulse_canon import PulseCanon
        from game.core.weapons.gatling_gun import GatlingGun
        from game.core.weapons.heavy_rail_gun import HeavyRailGun
        from game.core.weapons.plasma_canon import PlasmaCanon
        from game.core.weapons.homing_missile import HomingMissile
        from game.core.weapons.mine import Mine

        WeaponManager.weaponTypes = [
            PulseCanon,
            GatlingGun,
            HeavyRailGun,
            PlasmaCanon,
            HomingMissile,
            Mine,
        ]
        return WeaponManager.weaponTypes

    def nextWeapon(self):
        weaponTypes = WeaponManager._getWeaponTypes()
        ship = worldManager.findAttachedShip()
        currentWeaponId = (
            weaponTypes.index(ship.weapon) + 1
            if ship and ship.weapon in weaponTypes
            else 0
        )
        nextWeapon = (
            weaponTypes[currentWeaponId]
            if currentWeaponId < len(weaponTypes)
            else weaponTypes[0]
        )
        self.changeWeapon(nextWeapon)

    def previousWeapon(self):
        weaponTypes = WeaponManager._getWeaponTypes()
        ship = worldManager.findAttachedShip()
        currentWeaponId = (
            weaponTypes.index(ship.weapon) - 1
            if ship and ship.weapon in weaponTypes
            else -1
        )
        previousWeapon = (
            weaponTypes[currentWeaponId]
            if currentWeaponId >= 0
            else weaponTypes[-1]
        )
        self.changeWeapon(previousWeapon)

    @staticmethod
    def changeWeapon(weapon_idx_norm):
        weaponTypes = WeaponManager._getWeaponTypes()

        target_weapon = None
        if isinstance(weapon_idx_norm, type) and weapon_idx_norm in weaponTypes:
            target_weapon = weapon_idx_norm
        else:
            norm_value = float(np.clip(weapon_idx_norm, -1.0, 1.0))
            idx = int(np.round((norm_value + 1.0) / 2.0 * (len(weaponTypes) - 1)))
            target_weapon = weaponTypes[int(np.clip(idx, 0, len(weaponTypes) - 1))]

        from game.core.world.manager import worldManager
        ship = worldManager.findAttachedShip()
        if ship is not None:
            ship.weapon = target_weapon


weaponManager = WeaponManager()
